src/pykeen/models/unimodal/boxe_kg.py

```diff
@@ -0,0 +1,196 @@
+from typing import Any, ClassVar, Mapping, Optional
+
+from torch.nn.init import normal_, uniform_, zeros_
+from ...nn.modules import Interaction
+from ...losses import NSSALoss
+import torch
+from ...models import ERModel
+from ...nn.emb import EmbeddingSpecification
+from ...constants import DEFAULT_EMBEDDING_HPO_EMBEDDING_DIM_RANGE
+from ...typing import Hint, Initializer
+
+__all__ = [
+    'BoxEKG',
+]
+
+
+SANITY_EPSILON = 10 ** -8
+
+
+def product_normalise(input_tensor):   # Geometric normalize
+    step1_tensor = torch.abs(input_tensor)
+    step2_tensor = step1_tensor + SANITY_EPSILON    # Prevent zero values
+    log_norm_tensor = torch.log(step2_tensor)
+    step3_tensor = torch.mean(log_norm_tensor, dim=-1, keepdim=True)
+    norm_volume = torch.exp(step3_tensor)
+    pre_norm_out = input_tensor / norm_volume
+    return pre_norm_out
+
+
+def compute_box(base, delta, size):  # Compute box representations from embeddings
+    size_pos = torch.nn.functional.elu(size) + 1  # Enforce that sizes are strictly positive
+    delta_norm = product_normalise(delta)   # Shape vector is normalized
+    delta_final = torch.multiply(size_pos, delta_norm)   # Size is learned separately and applied
+    # Product normalize the delta
+    first_bound = base - 0.5 * delta_final
+    second_bound = base + 0.5 * delta_final
+    box_low = torch.minimum(first_bound, second_bound)
+    box_high = torch.maximum(first_bound, second_bound)
+    return box_low, box_high
+
+
+def dist(points, box_lows, box_highs):   # The dist function from the paper
```

Comment:
need type annotations for all inputs and return value, as well as a better name (note that there's a variable inside the function with the same name as the function - hugely problematic)